src/lesson_3.ts

```diff
@@ -0,0 +1,248 @@
+  private _grades: GradesType = {};
+  private _visits: VisitsType[] = [];
+
+  constructor(firstName: string, lastName: string, birthYear: number) {
+    this._firstName = firstName;
+    this._lastName = lastName;
+    this._birthYear = birthYear;
+  }
+
+  get getStudentFirstName(): string {
+    return this._firstName;
+  }
+
+  get getStudentLastName(): string {
+    return this._lastName;
+  }
+
+  get getStudentBirthYear(): number {
+    return this._birthYear;
+  }
+
+  get getStudentFullName(): string {
+    return `${this._lastName} ${this._firstName}`;
+  }
+
+  set setStudentFullName(value: string) {
+    [this._lastName, this._firstName] = value.split(" ");
+  }
+
+  get getStudentAge(): number {
+    return new Date().getFullYear() - this._birthYear;
+  }
+
+  set setStudentGrade(value: GradesType) {
+    this._grades = { ...this._grades, ...value };
+  }
+
+  set setStudentVisits(value: VisitsType[]) {
+    this._visits = [...this._visits, ...value];
+  }
+
+  getStudentPerformanceRating(): number {
+    const gradeValues: number[] = Object.values(this._grades);
+
+    if (!gradeValues.length) return 0;
+
+    const averageGrade: number =
+      gradeValues.reduce((sum: number, grade: number) => sum + grade, 0) /
+      gradeValues.length;
+    const attendancePercentage: number =
+      (this._visits.filter((present) => present).length / this._visits.length) *
+      100;
+
+    return (averageGrade + attendancePercentage) / 2;
+  }
+}
+
+const student = new Student("Bill", "Bob", 1920);
+type S = typeof student;
+
+const enum Statuses {
+  STATUS_1,
+  STATUS_2,
+  STATUS_3,
+}
+
+class Group<A> {
+  private _area: A | undefined;
+  private _status: Statuses | undefined;
+  private _id: string;
+
+  private _students: S[] = [];
+
+  directionName: string = "";
+  levelName: string = "";
+
+  constructor(directionName: string, levelName: string, id: string) {
+    this.directionName = directionName;
+    this.levelName = levelName;
+    this._id = id;
+  }
+
+  showPerformance() {
+    const sortedStudents = this._students.sort(
+      (a, b) =>
+        b.getStudentPerformanceRating() - a.getStudentPerformanceRating()
+    );
+    return sortedStudents;
+  }
+
+  set addStudent(value: S) {
+    this._students = [...this._students, value];
+  }
+
+  set removeStudent(value: S) {
+    this._students = this._students.filter(
+      (student) =>
+        student.getStudentAge !== value.getStudentAge &&
+        student.getStudentFullName !== value.getStudentFullName &&
+        student.getStudentAge != student.getStudentAge
+    );
+  }
+
+  get getGroupId(): string {
+    return this._id;
+  }
+
+  get getGroupArea(): A | undefined {
+    return this._area;
+  }
+
+  set setGroupArea(value: A) {
+    this._area = value;
+  }
+
+  get getGroupStatus(): Statuses | undefined {
+    return this._status;
+  }
+
+  set setGroupStatus(value: Statuses) {
+    this._status = value;
+  }
+}
+
+class Level<G extends Group<any>> {
```

Comment:
The Level class requirements were to implement getters for fields and add/remove group methods (addGroup / removeGroup). I don't see these methods in your class. Please be careful with the task requirements.

There is a good design principle - YAGNI principle ("You Aren't Gonna Need It") -  is a practice in software development which states that features should only be added when required.

It's okay to practice more than what's required in the task, as long as the task requirements are covered.